design_notes/XDN11 - Error handling.md

```diff
@@ -0,0 +1,345 @@
+failures. Just like with any other failure, developers will want to diagnose where and why the
+failure occurred. To assist with this, xlang will provide the ability for a projection to associate
+diagnostic information with the error. The following information can be associated:
+
+- Language specific error (i.e. error code, exception type)
+- Error message
+- Stack trace
+- Projection identifier
+- Any additional diagnostic information
+
+Some of this information can be represented by different languages in different ways. For instance,
+in some languages, stack traces can be an array of addresses while in others it can be a list of
+function names. Due to this, xlang will store the information generically as strings and it will
+be the responsibility of the projections to convert the representation it has to a string representation
+that can be logged and is readable by a developer or tooling.
+
+Errors may also end up traveling through several different components written in different languages
+before becoming a failure. To assist with diagnosing these failures, xlang will allow a projection to
+associate additional diagnostic information to the existing information as an error travels through it.
+This will help a developer get a complete picture of the failure which wouldn't be available with just
+the information collected from the language that initially returned the error. An example of this is
+the stack trace. A language will typically only be aware of the stack within its language boundary and
+thereby the diagnostic information collected at the initial error would only have a partial stack.
+The following additional information can be associated with the error as it travels through projections:
+
+- Stack trace
+- Projection identifier
+- Any additional diagnostic information
+
+The diagnostic information will be stored in a COM object (error info) implementing the
+IXlangErrorInfo interface. As with any COM object, it will be reference counted and will implement
+IUnknown. This will allow it to be easily accessed by projections written in different languages
+while still being able to ensure the cleanup of it.
+
+## Handling language errors
+
+Producing projections may encounter a language error when projecting an API call to a component.
+The error will either be from the translation of the call to the component's language or be from
+the component itself as a result of the call. Either way, upon encountering an error, the projection
+should determine whether the error is new or is a propagation of a previous xlang error. Strategies
+on doing this will be discussed in a later section.
+
+If the error is new, the projection should translate the error to an equivalent xlang error or to
+xlang_fail if there is no equivalent mapping. After translating the error, the projection should
+call xlang's error origination API, xlang_originate_error, to store the diagnostic information
```

Comment:
I feel like this should be called out earlier as a top level thing so that readers are familiar with all the pieces of the picture.

Reply:
Maybe we should have a diagram that shows end to end in the overview that includes error origination mention in passing, with details in this section.

Reply:
Addressed by talking about this in the overview and in the guidance to projections section.  Will also include a diagram.